hypothesis-python/docs/tutorial/flaky.rst

```diff
@@ -0,0 +1,107 @@
+Flaky failures
+==============
+
+Have you ever had a test fail, and then you re-run it, only for the test to magically pass? That is a *flaky test*. A flaky test is one which might behave differently when called again. You can think of it as a test which is not deterministic.
+
+Any test can be flaky, but because Hypothesis runs your test many times, Hypothesis tests are particularly likely to uncover flaky behavior.
+
+Note that Hypothesis does not require tests to be fully deterministic. Only the sequence of calls to Hypothesis APIs like ``draw`` from |st.composite| and the outcome of the test (pass or fail) need to be deterministic. This means you can use randomness, threads, or nondeterminism in your test, as long as it doesn't impact anything Hypothesis can see. We jokingly call this requirement "morally deterministic".
```

Comment:
I'd delete the morally-deterministic quip, it's a good one especially when talking to FP people about pure functions, but a little too niche for a tutorial 😢

Reply:
I'm ambivalent, I like it and think it's clear even to non-FP folks, but I might be too far away from a novice at this point to write a good tutorial. I removed it